### PR TITLE
cmd/swarm: pss smoke tests

### PR DESCRIPTION
**Pull request by skylenet**
Added smoke tests for raw, symmetric and asymmetric encrypted messages between random nodes.

Reviewed @ https://github.com/ethersphere/go-ethereum/pull/1305

### CODE REVIEW

File: cmd/swarm/swarm-smoke-pss/pss.go [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+}
+
+func (s *pssSession) sendRawMessage(sender *pssNode, receiver *pssNode, msg []byte) error {
+	return sender.client.Call(nil, "pss_sendRaw", hexutil.Encode(receiver.addr), s.topic, hexutil.Encode(msg))
+}
+
+func (s *pssSession) sendAsymMessage(sender *pssNode, receiver *pssNode, msg []byte) error {
+
+	// share public keys between nodes
+	err := sender.client.Call(nil, "pss_setPeerPublicKey", receiver.pubkey, s.topic, hexutil.Encode(receiver.addr))
```
Comment (nolash) [inline cmd/swarm/swarm-smoke-pss/pss.go]: Strictly speaking you don't have to do this every time you send an asym message. You only add to address book once.

File: cmd/swarm/swarm-smoke-pss/pss.go [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+			err = s.sendAsymMessage(j.sender, j.receiver, j.msg)
+			if err != nil {
+				log.Error("error sending asym message", "err", err)
+			}
+		default:
+			err = fmt.Errorf("invalid pssMode %d", mode)
+			log.Error("error processing job type", "err", err)
+		}
+
+		err = s.waitForMsg()
```
Comment (nolash) [inline cmd/swarm/swarm-smoke-pss/pss.go]: Hmm, I would prefer the send and checks were asynchronous operations. When done like this there is no effective concurrency.

File: cmd/swarm/swarm-smoke-pss/pss.go [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+
+	select {
+	case err := <-errc:
+		if err != nil {
+			metrics.GetOrRegisterCounter(fmt.Sprintf("pss.%s.fail", mode), nil).Inc(1)
+		}
+		return err
+	case <-time.After(time.Duration(timeout) * time.Second):
+		metrics.GetOrRegisterCounter(fmt.Sprintf("pss.%s.timeout", mode), nil).Inc(1)
+		return fmt.Errorf("timeout after %v sec", timeout)
```
Comment (nolash) [inline cmd/swarm/swarm-smoke-pss/pss.go]: Will this simple return free up all resources?

File: cmd/swarm/swarm-smoke-pss/pss.go (hunk 2 above)
Comment (skylenet) [reply cmd/swarm/swarm-smoke-pss/pss.go]: I was thinking about making both possible. Leaving it synchronous for now, and then adding a new PR to add a flag where you can run the same tests in an asynchronous mode. What do you think?

File: cmd/swarm/swarm-smoke-pss/main.go [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+			Action: wrapCliCommand("raw", pssRawCheck),
+		},
+		{
+			Name:   "all",
+			Usage:  "send and receive raw, sym and asym messages across randome nodes",
+			Action: wrapCliCommand("all", pssAllCheck),
+		},
+	}
+
+	sort.Sort(cli.FlagsByName(app.Flags))
```
[the remark this answers is not in the document]
Comment (skylenet) [reply cmd/swarm/swarm-smoke-pss/main.go]: It's nice to have the flags and commands ordered when running the `help` command.

File: cmd/swarm/swarm-smoke-pss/pss.go (hunk 1 above)
Comment (skylenet) [reply cmd/swarm/swarm-smoke-pss/pss.go]: I've added a map to track the known pubkeys. Now I'm only calling `setPeerPublicKey` once.

File: cmd/swarm/swarm-smoke-pss/pss.go (hunk 3 above)
Comment (skylenet) [reply cmd/swarm/swarm-smoke-pss/pss.go]: I think so. We have the following within the same function: 
```go
	defer func() {
		for _, n := range session.nodes {
			n.deregisterFunc()
		}
	}()
```

`deregisterFunc()` will close all connections/subscriptions.

**Comment by skylenet** [comment]
@nolash thanks for the review! I think I've fixed most of the things. Let me know if you have any other recommendations or concerns.

File: cmd/swarm/swarm-smoke-pss/pss.go (hunk 2 above)
Comment (nolash) [reply cmd/swarm/swarm-smoke-pss/pss.go]: It depends. Do you think the synchronous test is actually useful? I'm not so sure

**Comment by nonsense** [comment]
Please submit this to https://github.com/ethersphere/swarm since the Swarm codebase now lives there, and not under the `ethereum/go-ethereum` repo.

**Pull request closed without merging**